controllers/controller.go

```diff
@@ -0,0 +1,192 @@
+func (r *StatefulSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx).WithValues("statefulset", req.NamespacedName)
+
+	// Fetch StatefulSet of the request
+	old := appsv1.StatefulSet{}
+	if err := r.Get(ctx, req.NamespacedName, &old); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	sts := *old.DeepCopy()
+
+	// Get PVC information for all PVC that need to be resized
+	pvcs, err := r.getResizablePVCInfo(ctx, sts)
+	if err != nil {
+		l.Error(err, "Failed to get information of PVCs")
+		return ctrl.Result{}, err
+	}
+	if len(pvcs) == 0 {
+		// There are no PVCs that need to be resized.
+		// We do not need to handle this.
+		return ctrl.Result{}, nil
+	}
+	if sts.Labels != nil && sts.Labels[FailedLabel] == "true" {
+		// If this label is set the Sts needs human interaction, we cannot fix this.
+		// Common example for that is if the data transfer failed repeatedly, or if someone or something messed with the
+		// controller state and we are unable to recover.
+		return ctrl.Result{}, nil
+	}
+
+	// Perform resizing of all relevant PVCs of the StatefulSet, this includes the save scale down and scale up of the StatefulSet.
+	// This will not run through in one go but is meant to be called repeatedly until it succeeds.
+	// The resizing is idempotent and will return an errInProgress if we need to wait for asynchronous actions to complete.
+	res := ctrl.Result{}
+	sts, err = r.resize(ctx, sts, pvcs)
+	switch {
+	case errors.Is(err, errInProgress):
+		// Resizing is in progress, backing off.
+		res = ctrl.Result{
+			RequeueAfter: 5 * time.Second,
+		}
+	case errors.Is(err, errAbort):
+		// Resizing failed and cannot be recovered automatically.
+		// We can abort and scale up.
+		// We potentially ignore a critical error here, but as we already notify ops there is not much else to do.
+		sts, _ = scaleUp(sts)
+		fallthrough
+	case errors.Is(err, errCritical):
+		// Something went very wrong.
+		r.Recorder.Event(&sts, "Warning", "ErrorResize", err.Error())
+		l.Error(err, "Unable to resize PVCs and cannot recover")
+		if sts.Labels == nil {
+			sts.Labels = map[string]string{}
+		}
+		sts.Labels[FailedLabel] = "true"
+	case err == nil:
+		// We ran through successfully, so the resizing is complete.
+		// Cleanup annotation with PVCInfo so we do not try to resize again.
+		r.Recorder.Event(&sts, "Normal", "ResizeComplete", "Successfully resized StatefulSet")
+		l.Info("Successfully resized StatefulSet")
+		delete(sts.Annotations, PvcAnnotation)
+	default:
+		// Some potentially recoverable error. We will just back off exponentially and try again.
+		l.Error(err, "Unable to resize PVCs")
+		return ctrl.Result{}, err
+	}
+
+	// Apply possible changes from the resize function.
+	if !reflect.DeepEqual(sts.Annotations, old.Annotations) || !reflect.DeepEqual(sts.Spec, old.Spec) {
```

Comment:
Here you are comparing annotations and the spec, but not the labels. In line 116 the labels are updated, but I cannot see a subsequent `client.update` out of that change. The label change would be lost